adapters/Vungle/VungleAdapter/OpenBidding/GADMediationVungleInterstitial.m

```diff
@@ -0,0 +1,193 @@
+@end
+
+@implementation GADMediationVungleInterstitial {
+  /// Ad configuration for the ad to be loaded.
+  GADMediationInterstitialAdConfiguration *_adConfiguration;
+
+  /// The completion handler to call when an ad loads successfully or fails.
+  GADMediationInterstitialLoadCompletionHandler _adLoadCompletionHandler;
+
+  /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
+  id<GADMediationInterstitialAdEventDelegate> _delegate;
+
+  /// Indicates whether an interstitial ad is loaded.
+  BOOL _isAdLoaded;
+}
+
+@synthesize desiredPlacement;
+
+#pragma mark - GADMediationVungleInterstitial Methods
+
+- (nonnull instancetype)initWithAdConfiguration:(nonnull GADMediationInterstitialAdConfiguration*)adConfiguration
+                              completionHandler:(nonnull GADMediationInterstitialLoadCompletionHandler)completionHandler {
+  self = [super init];
+  if (self) {
+    _adConfiguration = adConfiguration;
+    self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:adConfiguration.credentials.settings networkExtras:adConfiguration.extras];
+
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationInterstitialLoadCompletionHandler origAdLoadHandler = [completionHandler copy];
+
+    /// Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationInterstitialAdEventDelegate>(
+      id<GADMediationInterstitialAd> ad, NSError *error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationInterstitialAdEventDelegate> delegate = nil;
+      if (origAdLoadHandler) {
+        delegate = origAdLoadHandler(ad, error);
+      }
+      origAdLoadHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestInterstitialAd {
+  if (!self.desiredPlacement.length) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(
+      GADMAdapterVungleErrorInvalidServerParameters,
+      @"Placement ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+
+  if ([[GADMAdapterVungleRouter sharedInstance] hasDelegateForPlacementID:self.desiredPlacement]) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(
+        GADMAdapterVungleErrorAdAlreadyLoaded,
+        @"Only a maximum of one ad per placement can be requested from Vungle.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+
+  if ([[GADMAdapterVungleRouter sharedInstance] isSDKInitialized]) {
+    [self loadAd];
+    return;
+  }
+
+  NSString *appID = [GADMAdapterVungleUtils findAppID:_adConfiguration.credentials.settings];
+  if (!appID) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(
+        GADMAdapterVungleErrorInvalidServerParameters, @"Vungle app ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
```

Comment:
Question: Shouldn't we check the app ID before we do the `loadAd` above?

In addition, it seems like these few lines are also implemented in `GADMediationVungleBanner.m` and `GADMediationVungleRewardedAd.m`, perhaps we can have a standalone method for in `GADMediationAdapterVungle.m` and implement it in `loadInterstitialForAdConfiguration`, e.t.c.?

Reply:
> Shouldn't we check the app ID before we do the loadAd above?
The app ID does not need to be checked in the loadAd above because the Vungle SDK has already initialized successfully. The app ID is only necessary for initialization and this section is to initialize the Vungle SDK if it hasn't been initialized yet.

> In addition, it seems like these few lines are also implemented in GADMediationVungleBanner.m and GADMediationVungleRewardedAd.m, perhaps we can have a standalone method in GADMediationAdapterVungle.m and implement it in loadInterstitialForAdConfiguration, e.t.c.?
These few lines are implemented in the banner and rewarded classes as well, but the `_adLoadCompletionHandler` is a different type. Perhaps a lazy initialization method should be added to the VungleRouter and have the delegate to handle failure passed in.

This case would be hit if the publisher does not call `[[GADMobileAds sharedInstance] startWithCompletionHandler:` and attempts to load an ad.